File: apps/server/src/modules/auth/auth.controller.ts
import type { Request, Response } from 'express';
import { authService } from './auth.service.js';
import { AUTH_MESSAGES } from './auth.constants.js';
import { asyncHandler } from '../../utils/asyncHandler.js';
import { ok, created } from '../../utils/ApiResponse.js';
import { ApiError } from '../../utils/ApiError.js';

function ctx(req: Request) {
  return { ipAddress: req.ip, userAgent: req.headers['user-agent'] };
}

export const authController = {
  register: asyncHandler(async (req: Request, res: Response) => {
    const result = await authService.register(req.body, ctx(req));
    return created(res, result, AUTH_MESSAGES.REGISTERED);
  }),

  login: asyncHandler(async (req: Request, res: Response) => {
    const result = await authService.login(req.body, ctx(req));
    return ok(res, result, AUTH_MESSAGES.LOGGED_IN);
  }),

  refresh: asyncHandler(async (req: Request, res: Response) => {
    const result = await authService.refresh(req.body.refreshToken, ctx(req));
    return ok(res, result, AUTH_MESSAGES.TOKEN_REFRESHED);
  }),

  logout: asyncHandler(async (req: Request, res: Response) => {
    await authService.logout(req.body.refreshToken ?? '');
    return ok(res, null, AUTH_MESSAGES.LOGGED_OUT);
  }),

  me: asyncHandler(async (req: Request, res: Response) => {
    const user = await authService.getMe(req.user!.id);
    return ok(res, user, 'Current user');
  }),

  updateProfile: asyncHandler(async (req: Request, res: Response) => {
    const user = await authService.updateProfile(req.user!.id, req.body);
    return ok(res, user, AUTH_MESSAGES.PROFILE_UPDATED);
  }),

  updateLocation: asyncHandler(async (req: Request, res: Response) => {
    const user = await authService.updateLocation(req.user!.id, req.body);
    return ok(res, user, AUTH_MESSAGES.PROFILE_UPDATED);
  }),

  changePassword: asyncHandler(async (req: Request, res: Response) => {
    const result = await authService.changePassword(req.user!.id, req.body);
    return ok(res, null, result.message);
  }),

  uploadAvatar: asyncHandler(async (req: Request, res: Response) => {
    const file = req.file;
    if (!file) throw ApiError.badRequest('Avatar image is required');
    const user = await authService.uploadAvatar(req.user!.id, {
      path: file.path,
      filename: file.filename,
      originalname: file.originalname,
      mimetype: file.mimetype,
      size: file.size,
    });
    return ok(res, user, AUTH_MESSAGES.AVATAR_UPDATED);
  }),

  forgotPassword: asyncHandler(async (req: Request, res: Response) => {
    const result = await authService.forgotPassword(req.body);
    return ok(res, null, result.message);
  }),

  resetPassword: asyncHandler(async (req: Request, res: Response) => {
    const result = await authService.resetPassword(req.body);
    return ok(res, null, result.message);
  }),
};

File: apps/guzo-api/src/main/java/et/guzo/web/controller/AuthController.java
package et.guzo.web.controller;

import et.guzo.common.ApiResponse;
import et.guzo.security.SecurityUtil;
import et.guzo.service.AuthService;
import et.guzo.web.dto.*;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/auth")
@RequiredArgsConstructor
public class AuthController {

    private final AuthService authService;

    @PostMapping("/register")
    public ApiResponse<LoginResponse> register(@Valid @RequestBody RegisterRequest body, jakarta.servlet.http.HttpServletRequest req) {
        return ApiResponse.ok(authService.register(body, req.getHeader("User-Agent"), req.getRemoteAddr()), "Registered");
    }

    @PostMapping("/login")
    public ApiResponse<LoginResponse> login(@Valid @RequestBody LoginRequest body, jakarta.servlet.http.HttpServletRequest req) {
        return ApiResponse.ok(authService.login(body, req.getHeader("User-Agent"), req.getRemoteAddr()), "Logged in");
    }

    @PostMapping("/refresh")
    public ApiResponse<RefreshResponse> refresh(@Valid @RequestBody RefreshRequest body, jakarta.servlet.http.HttpServletRequest req) {
        return ApiResponse.ok(authService.refresh(body.refreshToken(), req.getHeader("User-Agent"), req.getRemoteAddr()), "Token refreshed");
    }

    @PostMapping("/logout")
    public ApiResponse<Void> logout(@RequestBody(required = false) RefreshRequest body) {
        if (body != null) authService.logout(body.refreshToken());
        return ApiResponse.ok(null, "Logged out");
    }

    @PostMapping("/forgot-password")
    public ApiResponse<Map<String, String>> forgotPassword(@RequestBody ForgotPasswordRequest body) {
        return ApiResponse.ok(authService.forgotPassword(body));
    }

    @PostMapping("/reset-password")
    public ApiResponse<Map<String, String>> resetPassword(@Valid @RequestBody ResetPasswordRequest body) {
        return ApiResponse.ok(authService.resetPassword(body));
    }

    @GetMapping("/me")
    public ApiResponse<UserProfileDto> me() {
        return ApiResponse.ok(authService.me(SecurityUtil.requireUser().getId()));
    }

    @PatchMapping("/me")
    public ApiResponse<UserProfileDto> updateProfile(@RequestBody UpdateProfileRequest body) {
        return ApiResponse.ok(authService.updateProfile(SecurityUtil.requireUser().getId(), body));
    }

    @PatchMapping("/me/location")
    public ApiResponse<UserProfileDto> updateLocation(@RequestBody UpdateLocationRequest body) {
        return ApiResponse.ok(authService.updateLocation(SecurityUtil.requireUser().getId(), body));
    }

    @PatchMapping("/me/password")
    public ApiResponse<Void> changePassword(@Valid @RequestBody ChangePasswordRequest body) {
        authService.changePassword(SecurityUtil.requireUser().getId(), body);
        return ApiResponse.ok(null, "Password updated");
    }

    @PostMapping("/me/avatar")
    public ApiResponse<UserProfileDto> uploadAvatar(@RequestParam("avatar") org.springframework.web.multipart.MultipartFile file) {
        return ApiResponse.ok(authService.uploadAvatar(SecurityUtil.requireUser().getId(), file), "Avatar updated");
    }
}
